android/src/main/java/com/oney/WebRTCModule/WebRTCModule.java

```diff
@@ -38,45 +40,83 @@
+
+        public void setVideoDecoderFactory(VideoDecoderFactory videoDecoderFactory) {
+            this.videoDecoderFactory = videoDecoderFactory;
+        }
+
+        public void setVideoEncoderFactory(VideoEncoderFactory videoEncoderFactory) {
+            this.videoEncoderFactory = videoEncoderFactory;
+        }
+    }
+
     public WebRTCModule(ReactApplicationContext reactContext) {
+        this(reactContext, null);
+    }
+
+    public WebRTCModule(ReactApplicationContext reactContext, Options options) {
         super(reactContext);
 
         mPeerConnectionObservers = new SparseArray<>();
         localStreams = new HashMap<>();
 
-        ThreadUtils.runOnExecutor(() -> initAsync());
+        ThreadUtils.runOnExecutor(() -> initAsync(options));
     }
 
     /**
      * Invoked asynchronously to initialize this {@code WebRTCModule} instance.
      */
-    private void initAsync() {
+    private void initAsync(Options options) {
         ReactApplicationContext reactContext = getReactApplicationContext();
 
-        // Initialize EGL context required for HW acceleration.
-        EglBase.Context eglContext = EglUtils.getRootEglBaseContext();
-
         PeerConnectionFactory.initialize(
             PeerConnectionFactory.InitializationOptions.builder(reactContext)
                 .createInitializationOptions());
 
-        VideoEncoderFactory encoderFactory;
-        VideoDecoderFactory decoderFactory;
+        AudioDeviceModule adm = null;
+        VideoEncoderFactory encoderFactory = null;
+        VideoDecoderFactory decoderFactory = null;
+
+        if (options != null) {
+            adm = options.audioDeviceModule;
+            encoderFactory = options.videoEncoderFactory;
+            decoderFactory = options.videoDecoderFactory;
+        }
+
+        if (encoderFactory == null || decoderFactory == null) {
+            // Initialize EGL context required for HW acceleration.
+            EglBase.Context eglContext = EglUtils.getRootEglBaseContext();
 
-        if (eglContext != null) {
-            encoderFactory
-                = new DefaultVideoEncoderFactory(
+            if (eglContext != null) {
+                encoderFactory
+                    = new DefaultVideoEncoderFactory(
                     eglContext,
                     /* enableIntelVp8Encoder */ true,
                     /* enableH264HighProfile */ false);
-            decoderFactory = new DefaultVideoDecoderFactory(eglContext);
-        } else {
-            encoderFactory = new SoftwareVideoEncoderFactory();
-            decoderFactory = new SoftwareVideoDecoderFactory();
+                decoderFactory = new DefaultVideoDecoderFactory(eglContext);
+            } else {
+                encoderFactory = new SoftwareVideoEncoderFactory();
+                decoderFactory = new SoftwareVideoDecoderFactory();
+            }
+        }
+
+        if (adm == null) {
+            adm = JavaAudioDeviceModule.builder(reactContext).createAudioDeviceModule();
```

Comment:
Is that done by default by WebRTC ? I would probably prefer to let WebRTC do what it does if `adm` is not specified and left it not defined

Reply:
The answer is yes and no :-) At the moment, one is being created in the C++ layer, which matches this one, but that will be removed and one will need to be explicit about it.

I decided to be explicit now because I was confused when trying to understand what was going on, so this way people won't be confused when looking at it now.